src/errors.rs

```diff
@@ -0,0 +1,73 @@
+use crate::de;
+use crate::ser;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+/// Error type returned from the library
+pub enum AvroError {
+    /// Represents all cases of `std::io::Error`
+    #[error(transparent)]
+    IOError(#[from] std::io::Error),
+
+    /// Describes errors happened while decoding Avro data (except for `std::io::Error`)
+    #[error("Decoding error: {0}")]
+    DecodeError(String),
+
+    /// Describes errors happened while parsing Avro schemas
+    #[error("Failed to parse schema: {0}")]
+    ParseSchemaError(String),
+
+    /// Describes errors happened while performing schema resolution on Avro data.
+    #[error("Schema resolution error: {0}")]
+    SchemaResolutionError(String),
+
+    /// Describes errors happened while validating Avro data.
+    #[error("Validation error: {0}")]
+    ValidationError(String),
+
+    // TODO: figure out how to move the implementation of this error here
+    /// Describes errors that could be encountered while serializing data
+    #[error(transparent)]
+    SerError(#[from] ser::Error),
+
+    // TODO: figure out how to move the implementation of this error here
+    /// Describes errors that could be encountered while serializing data
+    #[error(transparent)]
+    DeError(#[from] de::Error),
+
+    // TODO: merge with SerError somehow?
+    // /// Represents all cases of `serde::ser::Error`
+    //#[error(transparent)]
+    //SerdeSerError(#[from] serde::ser::Error),
+    /// Describes errors happened trying to allocate too many bytes
+    #[error("Unable to allocate {desired} bytes (maximum allowed: {maximum})")]
+    MemoryAllocationError { desired: usize, maximum: usize },
+
+    /// Represents all cases of `uuid::Error`
+    #[error(transparent)]
+    UuidError(#[from] uuid::Error),
+
+    /// Describe a specific error happening with decimal representation
+    #[error("Number of bytes requested for decimal sign extension {requested} is less than the number of bytes needed to decode {needed}")]
+    SignExtendError { requested: usize, needed: usize },
+
+    // TODO: Should this be wrapped by ParseSchemaError somehow?
+    /// Represents all cases of `std::num::TryFromIntError`
+    #[error(transparent)]
+    TryFromIntError(#[from] std::num::TryFromIntError),
+
+    // TODO: Should this be wrapped by ParseSchemaError somehow?
+    /// Represents all cases of `serde_json::Error`
+    #[error(transparent)]
+    JSONError(#[from] serde_json::Error),
+
+    // TODO: Should this be wrapped by SchemaResolutionError somehow?
```

Comment:
Just an observation but I think that many of these `TODO`s go away if you make more specific error variants.

Reply:
I think you are right. I think I missed the fact that having many errors is not a problem in Rust. Although, I am curious to see at what point I should go for the approach that @athre0z suggested above:
>If you have many many failure cases I'd even recommend splitting them up into multiple different error types, e.g. DecodingError, NetworkError that then can all be into'ed a common error enum type with branches for each more granular error type. This allows for being more verbose in describing what exactly went wrong without forcing users to match on an unreasonable amount of error types when they actually just want to catch e.g. a decoding error in general.